src/dbnode/storage/series/buffer.go

```diff
@@ -1152,11 +1153,16 @@ func (b *BufferBucket) write(
 	for i := range b.encoders {
 		lastWriteAt := b.encoders[i].lastWriteAt
 		if timestamp.Equal(lastWriteAt) {
-			last, err := b.encoders[i].encoder.LastEncoded()
+			lastDatapoint, err := b.encoders[i].encoder.LastEncoded()
 			if err != nil {
 				return false, err
 			}
-			if last.Value == value {
+			lastAnnotation, err := b.encoders[i].encoder.LastAnnotation()
+			if err != nil {
+				return false, err
+			}
+
+			if lastDatapoint.Value == value && bytes.Equal(lastAnnotation, annotation) {
```

Comment:
nit - This works but I think we make the abstraction cleaner -

Instead of exposing lastdatapoint and lastannotation from encoders which seems like is done only to check if datapoint is being repeated what if we expose another function called  

```
IsSameAsPreviousEncodedValue(dp ts.Datapoint, unit xtime.Unit, annotation ts.Annotation) bool 
```
(function name is only suggestive. Note that function signature is same as Encode() which makes this function signature easy to understand)

That way we would not have to modify this piece of code for any future encoders (which can have their own interpretation of datapoint and annotation) and we will also reduce the scope of introducing an unintended bug.

For eg. this boolean check only works if protobuf encoding sets datapoint value to 0. But protobuf encoding can very easily choose not to do so since its value is not integral to its encoding logic.

Reply:
Hmm I don't think the encoder level should expose this specific function. Although yes, `LastEncoded` and `LastAnnotation` is currently only used in this way, I think it should be the responsibility of the user of the encoder to figure this out themselves.

Exposing a `LastAnnotation` function is pretty future-proof, since the only data we receive from the user is in the value and annotation. Exposing these functions separately gives more flexibility to users of the encoder to do what they wish with it.